arch/riscv/kernel/irq.c

```diff
@@ -58,7 +58,7 @@ struct irq_chip riscv_irq_chip = {
 void __init init_IRQ(void)
 {
 	unsigned int irq;
-	for (irq = 0; irq < NR_IRQS; irq++)
+	for (irq = 0; irq < 2; irq++)
```

Comment:
> I like your suggestion, but is there something more specific we can use than "BASE"? These are the CPU interrupt numbers, if I understand the code correctly.

Right, the interrupts that are currently specified in the privileged ISA are those originating from the processor itself, plus the non-standard HTIF extension which is gradually being deprecated.  Hence, one possibility is NR_IRQS_CORE, although other suggestions are welcome, too.

> Which leads me to my next question, which is how does the CPU receive an interrupt from a peripheral device? That would be neither timer nor software interrupt.

The upcoming platform specification will introduce a global interrupt controller for the external interrupts.  Without delving into too many details, the intent is to support both the traditional driver model, where the kernel has direct access to devices, and a cleaner scheme where low-level drivers can be separated from the kernel and run in an isolated environment.  With the latter, having a leaner "paravirtualized" device interface would also help abstract platform differences so that the same kernel image can support a variety of systems.  In the extreme case, the kernel only sees software interrupts raised by H or M mode corresponding to these virtualized devices.

To quote a previous riscv-hw post of Krste's:

> We obviously designed the RISC-V privileged architecture around the
> idea of having virtualized devices be the standard way that a
> supervisor-level OS interacts with I/O hardware.  We intended this to
> support both separate off-load cores running the device drivers, as
> well as multiplexing the OS cores to run the devices (without OS
> knowing).  This is not the only way, and the bad old way is fully
> supported too.

> We do need to define a new Driver Execution Environment (DEE) and
> figure out how to protect the system form the driver, while cleanly
> interoperating with OS/Hypervisor memory translation schemes,
> including page faults, and possibly also stealing CPU cycles from
> OS/Hypervisor cores.